Add IOS Parser "show vlan internal usage"

cherifimehdi opened the pull request:
## Description
New IOS/IOSXE Parser "show vlan internal usage"

## Motivation and Context
I noticed that there is no parser for "show vlan internal usage" and this contribution will help network engineer 

## Impact (If any)
<!--- If there is any negative impact - what is it? -->

## Screenshots:
**Testing ShowVlanInternalUsage  class:**

![test1](https://github.com/CiscoTestAutomation/genieparser/assets/73710391/bf8fcfe5-cbb0-4054-baf3-66074e6218f6)


**Testing ShowVlanInternalUsage_iosxe  class:**

![test2](https://github.com/CiscoTestAutomation/genieparser/assets/73710391/decc9d34-9262-482d-a236-df544e47ab3f)


**Testing the "show vlan internal usage" parser:**

                                                                                       
![script](https://github.com/CiscoTestAutomation/genieparser/assets/73710391/7b0a7cef-2bb2-445c-9508-786065bff0e8)


![test_parser](https://github.com/CiscoTestAutomation/genieparser/assets/73710391/a85bb434-a51f-4c4c-b688-f2c419581f6e)




## Checklist:
<!--- This is meant more as a personal checklist so we don't forgot important steps! -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have updated the changelog.
- [x] I have updated the documentation (If applicable).
- [x] I have added tests to cover my changes (If applicable).
- [x] All new and existing tests passed.
- [x] All new code passed compilation.

cherifimehdi commented:
Hello, two days ago I added a new parser for IOS "show vlan internal usage"

cherifimehdi commented:
Hi, please have you checked my submission for a parser? Thank you

ThomasJRyan commented on src/genie/libs/parser/ios/show_vlan_internal_usage.py:
```diff
@@ -0,0 +1,65 @@
+        show vlan internal usage"""
+
+    schema = {
+        'Internal Vlan': {
```
Key names should all be lowercase with spaces replaced by underscores
```suggestion
        'internal_vlan': {
```

ThomasJRyan commented on src/genie/libs/parser/ios/show_vlan_internal_usage.py:
```diff
@@ -0,0 +1,65 @@
+    schema = {
+        'Internal Vlan': {
+            Any(): {
+                        'Usage': str,
```
```suggestion
                        'usage': str,
```

ThomasJRyan commented on src/genie/libs/parser/ios/show_vlan_internal_usage.py:
```diff
@@ -0,0 +1,65 @@
+# =================================================
+# Parser for 'show vlan internal usage' command
+# =================================================
+class ShowVlanInternalUsage(ShowVlanInternalUsageSchema):
```
This parser and its schema should be under `show_vlan.py`, rather than creating a new file for it. Additionally, for `ios` devices which are typically identical to `iosxe` devices, we'll usually put the parser under `iosxe` and import it into `ios`. You can see an example of what I mean in `src/genie/libs/parser/ios/show_vlan.py`

cherifimehdi replied:
please should I remove the file show_vlan_internal_usage.py and pull directly my change under show_vlan.py?

ThomasJRyan commented on src/genie/libs/parser/ios/show_vlan_internal_usage.py:
```diff
@@ -0,0 +1,65 @@
+
+            m = p0.match(line)
+            if m:
+                internal_vlan_dict=parsed_dict.setdefault('Internal Vlan', {})
+                Usage=m.groupdict()['Usage']
+                VLAN=m.groupdict()['VLAN']
+                internal_vlan_dict[VLAN] = {}
+                internal_vlan_dict[VLAN]['Usage']=Usage
```
Just a bit of cleaning up
```suggestion
                internal_vlan_dict = parsed_dict.setdefault('internal_vlan', {})
                usage = m.groupdict()['Usage']
                vlan = m.groupdict()['VLAN']
                vlan_dict = internal_vlan_dict[vlan].set_default(vlan, {})
                vlan_dict['usage'] = usage
```

ThomasJRyan reviewed (changes requested):
Sorry for the delay. There are a few things that need to be address in this PR, but overall it's looking pretty good

cherifimehdi commented:
Please shoud I continue with this PR or close it and open a new one to move the parser to show_vlan.py under IOSXE?

cherifimehdi commented:
Hello, I addressed all the changes asked for:
1. I deleted the file "show_vlan_internal_usage.py" under IOS parser and I moved it in "show_vlan.py" under IOSXE
2. I imported the parser "show_vlan_internal_usage" from IOSXE into IOS as asked under "show_vlan.py" under IOS
3. I deleted the folder "ShowVlanInternalUsage" under tests of IOS and moved it into tests of IOSXE
3. I did all the test as you can see it from the updated test with screenshots
Thank you for your review

cherifimehdi commented:
Hello
Please can you review my PR cause I addressed all the changes according to
your review
Thank you

Le ven. 24 nov. 2023 à 16:08, Thomas Ryan ***@***.***> a
écrit :

> ***@***.**** requested changes on this pull request.
>
> Sorry for the delay. There are a few things that need to be address in
> this PR, but overall it's looking pretty good
> ------------------------------
>
> In src/genie/libs/parser/ios/show_vlan_internal_usage.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/796#discussion_r1404444214>
> :
>
> > +import re
> +
> +# Metaparser
> +from genie.metaparser import MetaParser
> +from genie.metaparser.util.schemaengine import Any
> +
> +
> +# ====================================================
> +#  Schema for 'show vlan internal usage' command
> +# ====================================================
> +class ShowVlanInternalUsageSchema(MetaParser):
> +    """Schema for:
> +        show vlan internal usage"""
> +
> +    schema = {
> +        'Internal Vlan': {
>
> Key names should all be lowercase with spaces replaced by underscores
> ⬇️ Suggested change
>
> -        'Internal Vlan': {
> +        'internal_vlan': {
>
> ------------------------------
>
> In src/genie/libs/parser/ios/show_vlan_internal_usage.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/796#discussion_r1404444594>
> :
>
> > +# Metaparser
> +from genie.metaparser import MetaParser
> +from genie.metaparser.util.schemaengine import Any
> +
> +
> +# ====================================================
> +#  Schema for 'show vlan internal usage' command
> +# ====================================================
> +class ShowVlanInternalUsageSchema(MetaParser):
> +    """Schema for:
> +        show vlan internal usage"""
> +
> +    schema = {
> +        'Internal Vlan': {
> +            Any(): {
> +                        'Usage': str,
>
> ⬇️ Suggested change
>
> -                        'Usage': str,
> +                        'usage': str,
>
> ------------------------------
>
> In src/genie/libs/parser/ios/show_vlan_internal_usage.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/796#discussion_r1404446830>
> :
>
> > +class ShowVlanInternalUsageSchema(MetaParser):
> +    """Schema for:
> +        show vlan internal usage"""
> +
> +    schema = {
> +        'Internal Vlan': {
> +            Any(): {
> +                        'Usage': str,
> +            },
> +        }
> +    }
> +
> +# =================================================
> +# Parser for 'show vlan internal usage' command
> +# =================================================
> +class ShowVlanInternalUsage(ShowVlanInternalUsageSchema):
>
> This parser and its schema should be under show_vlan.py, rather than
> creating a new file for it. Additionally, for ios devices which are
> typically identical to iosxe devices, we'll usually put the parser under
> iosxe and import it into ios. You can see an example of what I mean in
> src/genie/libs/parser/ios/show_vlan.py
> ------------------------------
>
> In src/genie/libs/parser/ios/show_vlan_internal_usage.py
> <https://github.com/CiscoTestAutomation/genieparser/pull/796#discussion_r1404449331>
> :
>
> > +                internal_vlan_dict=parsed_dict.setdefault('Internal Vlan', {})
> +                Usage=m.groupdict()['Usage']
> +                VLAN=m.groupdict()['VLAN']
> +                internal_vlan_dict[VLAN] = {}
> +                internal_vlan_dict[VLAN]['Usage']=Usage
>
> Just a bit of cleaning up
> ⬇️ Suggested change
>
> -                internal_vlan_dict=parsed_dict.setdefault('Internal Vlan', {})
> -                Usage=m.groupdict()['Usage']
> -                VLAN=m.groupdict()['VLAN']
> -                internal_vlan_dict[VLAN] = {}
> -                internal_vlan_dict[VLAN]['Usage']=Usage
> +                internal_vlan_dict = parsed_dict.setdefault('internal_vlan', {})
> +                usage = m.groupdict()['Usage']
> +                vlan = m.groupdict()['VLAN']
> +                vlan_dict = internal_vlan_dict[vlan].set_default(vlan, {})
> +                vlan_dict['usage'] = usage
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CiscoTestAutomation/genieparser/pull/796#pullrequestreview-1748093924>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARSLWN67QSL6AWPDNKYPCIDYGCZ5NAVCNFSM6AAAAAA6ZXKTRGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONBYGA4TGOJSGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

tahigash commented:
> Hello, Thank you for reviewing and accepting my changes. Please see "Changes approved" by 02 reviewers,"All checks have passed" and "This branch has no conflicts with the base branch". Please, does this mean my parser has merged or this needs to be granted by the third reviewer? Thank you Le mar. 28 nov. 2023 à 17:32, Mehdi CHERIFI ***@***.***> a écrit :
> […](#)
> Hello Please can you review my PR cause I addressed all the changes according to your review Thank you Le ven. 24 nov. 2023 à 16:08, Thomas Ryan ***@***.***> a écrit : > ***@***.**** requested changes on this pull request. > > Sorry for the delay. There are a few things that need to be address in > this PR, but overall it's looking pretty good > ------------------------------ > > In src/genie/libs/parser/ios/show_vlan_internal_usage.py > <[#796 (comment)](https://github.com/CiscoTestAutomation/genieparser/pull/796#discussion_r1404444214)> > : > > > +import re > + > +# Metaparser > +from genie.metaparser import MetaParser > +from genie.metaparser.util.schemaengine import Any > + > + > +# ==================================================== > +# Schema for 'show vlan internal usage' command > +# ==================================================== > +class ShowVlanInternalUsageSchema(MetaParser): > + """Schema for: > + show vlan internal usage""" > + > + schema = { > + 'Internal Vlan': { > > Key names should all be lowercase with spaces replaced by underscores > ⬇️ Suggested change > > - 'Internal Vlan': { > + 'internal_vlan': { > > ------------------------------ > > In src/genie/libs/parser/ios/show_vlan_internal_usage.py > <[#796 (comment)](https://github.com/CiscoTestAutomation/genieparser/pull/796#discussion_r1404444594)> > : > > > +# Metaparser > +from genie.metaparser import MetaParser > +from genie.metaparser.util.schemaengine import Any > + > + > +# ==================================================== > +# Schema for 'show vlan internal usage' command > +# ==================================================== > +class ShowVlanInternalUsageSchema(MetaParser): > + """Schema for: > + show vlan internal usage""" > + > + schema = { > + 'Internal Vlan': { > + Any(): { > + 'Usage': str, > > ⬇️ Suggested change > > - 'Usage': str, > + 'usage': str, > > ------------------------------ > > In src/genie/libs/parser/ios/show_vlan_internal_usage.py > <[#796 (comment)](https://github.com/CiscoTestAutomation/genieparser/pull/796#discussion_r1404446830)> > : > > > +class ShowVlanInternalUsageSchema(MetaParser): > + """Schema for: > + show vlan internal usage""" > + > + schema = { > + 'Internal Vlan': { > + Any(): { > + 'Usage': str, > + }, > + } > + } > + > +# ================================================= > +# Parser for 'show vlan internal usage' command > +# ================================================= > +class ShowVlanInternalUsage(ShowVlanInternalUsageSchema): > > This parser and its schema should be under show_vlan.py, rather than > creating a new file for it. Additionally, for ios devices which are > typically identical to iosxe devices, we'll usually put the parser under > iosxe and import it into ios. You can see an example of what I mean in > src/genie/libs/parser/ios/show_vlan.py > ------------------------------ > > In src/genie/libs/parser/ios/show_vlan_internal_usage.py > <[#796 (comment)](https://github.com/CiscoTestAutomation/genieparser/pull/796#discussion_r1404449331)> > : > > > + internal_vlan_dict=parsed_dict.setdefault('Internal Vlan', {}) > + Usage=m.groupdict()['Usage'] > + VLAN=m.groupdict()['VLAN'] > + internal_vlan_dict[VLAN] = {} > + internal_vlan_dict[VLAN]['Usage']=Usage > > Just a bit of cleaning up > ⬇️ Suggested change > > - internal_vlan_dict=parsed_dict.setdefault('Internal Vlan', {}) > - Usage=m.groupdict()['Usage'] > - VLAN=m.groupdict()['VLAN'] > - internal_vlan_dict[VLAN] = {} > - internal_vlan_dict[VLAN]['Usage']=Usage > + internal_vlan_dict = parsed_dict.setdefault('internal_vlan', {}) > + usage = m.groupdict()['Usage'] > + vlan = m.groupdict()['VLAN'] > + vlan_dict = internal_vlan_dict[vlan].set_default(vlan, {}) > + vlan_dict['usage'] = usage > > — > Reply to this email directly, view it on GitHub > <[#796 (review)](https://github.com/CiscoTestAutomation/genieparser/pull/796#pullrequestreview-1748093924)>, > or unsubscribe > <https://github.com/notifications/unsubscribe-auth/ARSLWN67QSL6AWPDNKYPCIDYGCZ5NAVCNFSM6AAAAAA6ZXKTRGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONBYGA4TGOJSGQ> > . > You are receiving this because you authored the thread.Message ID: > ***@***.***> >

The merge is supposed to be done by pyATS team after having 2 approvals. 

@ThomasJRyan or @Lelor please merge if no issue for merging.

cherifimehdi commented:
Hi @ThomasJRyan  @Lelor  @sowmyadn010501  Does merging take time after changes approved by two reviewers and all test pass with no conflict with the base branch? Thank you

cherifimehdi commented:
@ThomasJRyan Thank you for merging my contribution, and thanks to @Lelor @tahigash
